### ISSUE TITLE
Unexpected error when passing generic user-defined type guard into generic function

### ISSUE THREAD

**Issue by OliverJAsh** [microsoft/TypeScript#45184]
# Bug Report

### 🔎 Search Terms

user defined type guard refinement 

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about generic functions and user-defined type guards

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/C4TwDgpgBAsiDiEB2EBOBLAxgVSeg9kgDwAqAfFALxQDeUwAhgOYBcUA5AILsDcUAbgwA2AVwhsSUAL5QAPrXrM27AEK8BwsROk8AULoAmETEIapomQgGdgUAGZI2pMgAoGbOIhQYceQs4BKKgoGKHQrBUZWDjU+QVFxKEkpPV0AejSoADl8KDRUfFQoBgiIAA9ITGAIAyhAUHJDY1NzKEskGyhzACtjYCdOVwc2NzZOIMoKACN8fCEIBiRxihc7dyhOAG0AXSX17b1u3pXF1KMTMwtrW0OqgCZ+wccoEfXd6dn5xeCoIhU8suqSAMEQGK0mbBU212kK2ejSACpdLhypVqrV8oUoIAZcl0nFQTBEAFtkLZ8HZ6OBoOxnC8AGpmdALYCBb6hcKRJRQABEKi5cU0iRIfCk7DCESQ+FsJSs6CYSAYkzm9FyYDMDGJ1SKZIpkA4LxESAA1hKAO5fCZQd5zBbsAB0uigSUpEW1qtQ6ogmoi7AYooWtR9fpa6CQlkJquA6EVEHtjsdJEpHANxvwZtF7IlUqsMrlCqVwFyoF17HpGCZRGTpqQZDtDrj8cT7ErqaQ6fFkuK2dl8ujyp1VLoUTYPL5GgSbGbZuFtfhaV0N2At2OAT0QA)

### 💻 Code

```ts
type MyGenericUnion<T> = { tag: 'A'; value: T } | { tag: 'B'; value: T };

declare const fn: <T>(a: MyGenericUnion<T>) => a is { tag: 'B'; value: T };

// No error as expected ✅
declare const reject: <A>(fn: (a: A) => boolean) => (fa: A[]) => A[];
reject(fn);

declare const reject2: <A>(fn: (a: A) => boolean) => <B extends A>(fb: B[]) => B[];
/*
Unexpected error ❌
Argument of type '<T>(a: Variant<T>) => a is { tag: "B"; value: T; }' is not assignable to parameter of type '(a: unknown) => boolean'.
  Types of parameters 'a' and 'a' are incompatible.
    Type 'unknown' is not assignable to type 'Variant<unknown>'.
      Type 'unknown' is not assignable to type '{ tag: "B"; value: unknown; }'.
*/
reject2(fn)
```

### 🙁 Actual behavior

See comments in code.

### 🙂 Expected behavior

See comments in code.

**Comment by RyanCavanaugh** [comment]
Here's an example that shows what's going on
```ts
  declare const fn: <T>(a: { value: T }) => void;

  // Differs only in return type
  function pass1<A>(fn: (x: A) => void): (fa: A) => void {
    return fn;
  }
  function pass2<A>(fn: (x: A) => void) {
    return fn;
  }
  function fail<A>(fn: (x: A) => void): <B extends A>(fb: B) => void {
    return fn;
  }

  // No error as expected ✅
  const p1 = pass1(fn);
  // Hidden type parameter exists on 'p1'!
  p1({ value: 3 });
  const p2 = pass2(fn);
  // Hidden type parameter exists on 'p2'!
  p2({ value: 3 });

  // Unexpected error ❌
  const f = fail(fn);
  // What type "should" f have?
```

In TypeScript 3.3.3, this fails, because we're doing contravariant-only inference to `A`, so the correct inference is `unknown`, and it would be unsound to return a function reference to a function taking `unknown` (effectively `any` in an input position) when we know it's actually bound to a function taking a more specific type (even if it's an unconstrained generic - consider the case where you have *two* references to the type parameter to see why).

However, as of TS 3.5, we can cheat our way out of this by preserving the genericness of the function by attaching a type parameter to the return type. This is what happens in `pass1` and `pass2` here. However, we can only attached type parameter lists where one doesn't exist already -- this is what causes `fail` to fail.

The only remaining weirdness here is that the `extends A` clause in `fail` isn't necessary to make this weirdness manifest. I'm not entirely sure how to reason about that -- if your type parameter only occurs in the parameter list in contravariant positions, it is truly not inferable, so the case is sort of moot.

cc @ahejlsberg in case I got any of this wrong

**Comment by RyanCavanaugh** [comment]
Sort of working-as-intended but I'll say design limitation since in theory we could manifest hidden type parameter lists (but won't)
